xapian-applications/omega/handler_gmime.cc

```diff
@@ -0,0 +1,183 @@
+	parser.reset();
+	parser.ignore_metarobots();
+	parser.parse_html(text, newcharset, true);
+	dump += parser.dump;
+    } catch (...) {
+	return;
+    }
+}
+
+size_t decode(unsigned char* text, size_t len, GMimeContentEncoding encoding)
+{
+    unsigned char buffer[SIZE];
+    int state;
+    guint32 save;
+    switch (encoding) {
+	case GMIME_CONTENT_ENCODING_BASE64: {
+	    len = g_mime_encoding_base64_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_UUENCODE: {
+	    len = g_mime_encoding_uudecode_step(text, len, buffer,
+						&state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	case GMIME_CONTENT_ENCODING_QUOTEDPRINTABLE: {
+	    len = g_mime_encoding_quoted_decode_step(text, len, buffer,
+						     &state, &save);
+	    memcpy(text, buffer, len);
+	    break;
+	}
+	default:
+	    break;
+    }
+    return len;
+}
+
+bool
+parser_content(GMimeObject* me, string& dump)
+{
+    GMimeContentType* ct = g_mime_object_get_content_type(me);
+    if (GMIME_IS_MULTIPART(me)) {
+	GMimeMultipart* mpart = (GMimeMultipart*)me;
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	int count = g_mime_multipart_get_count(mpart);
+	if (subtype == "alternative") {
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		if (parser_content(part, dump))
+		    return true;
+	    }
+	} else {
+	    bool ret = false;
+	    for (int i = 0; i < count; ++i) {
+		GMimeObject* part = g_mime_multipart_get_part(mpart, i);
+		ret |= parser_content(part, dump);
+	    }
+	    return ret;
+	}
+    } else if (GMIME_IS_PART(me)) {
+	GMimePart* part = (GMimePart*)me;
+	GMimeDataWrapper* content = g_mime_part_get_content_object(part);
+	string type = g_mime_content_type_get_media_type(ct);
+	string subtype = g_mime_content_type_get_media_subtype(ct);
+	string charset(g_mime_object_get_headers(me));
+	size_t end, start;
+	start = charset.find("charset=");
+	if (start != string::npos) {
+	    start += 8;
+	    if (charset[start] == '"')
+		start++;
+	    for (end = start; end < charset.length() &&
+		 !isspace(charset[end]) && charset[end] != '"'; ++end);
+	    charset = charset.substr(start, end - start);
+	    charset = g_mime_charset_canon_name(charset.c_str());
+	} else {
+	    charset.clear();
+	}
+	if (type == "text") {
+	    char buffer[SIZE];
+	    GMimeStream* sr = g_mime_data_wrapper_get_stream(content);
+	    size_t len = g_mime_stream_read(sr, buffer, SIZE);
+	    auto encoding = g_mime_part_get_content_encoding(part);
+	    len = decode((unsigned char*)buffer, len, encoding);
+	    if (subtype == "plain") {
+		if (!charset.empty() && charset != "UTF-8") {
+		    GMimeFilter* ch = g_mime_filter_charset_new(charset.c_str(),
+								"UTF-8");
+		    if (GMIME_IS_FILTER(ch)) {
+			size_t presize = 0;
+			g_mime_filter_complete(ch, buffer, len, presize,
+					      (char**)&buffer, &len, &presize);
```

Comment:
Here you are specifying the input buffer as the output buffer.  That's unlikely to work well if translating to UTF-8 from a charset if the output is longer than the input, because the conversion will start to overwrite the input before it has read it.

For example, consider iso-8859-1 input "República" which consists of the 9 bytes:

52 65 70 fa 62 6c 69 63 61

I'll assume the input is read character by character (it's possible there's some funky optimisation which means the convertor reads more data at a time, but the general point still holds and it's simpler to consider the straightforward case to show how this might go wrong).

So as that gets converted to UTF-8 things are OK for the first 3 characters, as they each map to a single byte UTF-8 sequence (which consists of the same byte value as the input, since ASCII is a subset of UTF-8).  But when we convert the 4th character we get the two byte sequence c3 ba and the buffer now looks like this:

52 65 70 c3 ba 6c 69 63 61

Then we try to convert the 5th character, and we read byte `ba` which has overwritten `62`.  That's converted to c2 ba in UTF-8 so now we have:

52 65 70 c3 c2 ba 69 63 61

And again we have `ba`, so repeating these steps to the end the final result would be this, which isn't even valid UTF-8:

52 65 70 c3 c2 c2 c2 c2 c2 ba

That ought to be:

52 65 70 c3 ba 62 6c 69 63 61

Another problem is that the input might fit in `SIZE` bytes but the output might not.

To use this function safely I think you need to give it a separate output buffer, and to size that buffer such that it's large enough for any possibly expansion that character set conversion can give (or to retry if the conversion doesn't complete).

However Omega already has a function to convert text to UTF-8, so probably it makes more sense to just use that since it's much simpler to use from C++ - see `utf8convert.h`:

```
void convert_to_utf8(std::string & text, const std::string & charset);
```

This takes a string by reference and the charset it is in and modifies the string to be in UTF-8.  It even handles checking if it's already UTF-8 for you (it doesn't currently check for us-ascii but I'll push a fix for that shortly as that's a missed optimisation opportunity).

Reply:
Okay, I has used that function before. I will change the filters for `convert_to_utf8`